Add Dockerfile and client.py; deploy to EC2 on AWS via Github Actions (work in progress)

abrichr opened the pull request:
## Instructions

1. Fork https://github.com/OpenAdaptAI/SoM, clone your fork to your local machine, and check out the feat/deploy branch.
2. Follow the instructions at the top of [deploy.py](https://github.com/OpenAdaptAI/SoM/blob/feat/deploy/deploy.py).

## Summary

This PR implements functionality to automatically deploy OmniParser to EC2 on AWS via Docker, Github Actions, and boto3.

It also adds a working example `client.py`.

TODO:
- update with screenshots like in https://github.com/microsoft/SoM/pull/19

abrichr commented:
@yadong-lu ready for review 🙏

abrichr commented:
@yadong-lu thank you for the excellent work on this repo!

I believe this PR is ready to merge. Please let me know if you have any questions/comments!

yadong-lu commented:
> @yadong-lu thank you for the excellent work on this repo!
> 
> I believe this PR is ready to merge. Please let me know if you have any questions/comments!

Thanks! Will review this soon.

abrichr commented:
@yadong-lu I am planning to refactor the common code between this and https://github.com/microsoft/SoM/pull/19 (e.g., deploy.py) into a separate module for importing, reducing boilerplate and keeping deployments consistent. 

The new `deploy.py` here would look like this:
```
"""
OmniParser Deployment Script

This script leverages OpenAdapter's Deploy interface to streamline deployment, setup, and management of the OmniParser server on AWS.

Setup Instructions:
    1. Create a `.env` file in the same directory with the following keys:

        AWS_ACCESS_KEY_ID=<your AWS access key>
        AWS_SECRET_ACCESS_KEY=<your AWS secret access key>
        AWS_REGION=<your AWS region>
        GITHUB_OWNER=<GitHub user or org>
        GITHUB_REPO=<repo name>
        GITHUB_TOKEN=<GitHub token>
        PROJECT_NAME=<AWS project tag>

    2. Install dependencies:

        pip install -r deploy_requirements.txt

Usage Examples:
    Start server:
        python deploy.py start

    Stop server and clean up resources:
        python deploy.py stop

    Check server status:
        python deploy.py status

    SSH into server:
        python deploy.py ssh
"""

from openadapter.server import OpenAdapterConfig, Deploy
import fire

config = OpenAdapterConfig(
    AWS_ACCESS_KEY_ID="",
    AWS_SECRET_ACCESS_KEY="",
    AWS_REGION="us-west-2",
    GITHUB_OWNER="OpenAdaptAI",
    GITHUB_REPO="OmniParser",
    GITHUB_TOKEN="",
    PROJECT_NAME="OmniParser"
)

class OmniParserDeploy(Deploy):
    def __init__(self):
        super().__init__(config=config)

if __name__ == "__main__":
    fire.Fire(OmniParserDeploy)
```

Let me know if you'd like to wait until this new module is ready before merging, or if you'd prefer proceeding now with the current structure. Thanks again for your great work on this repo!

abrichr commented:
@yadong-lu I am planning to refactor the common code between this and https://github.com/microsoft/SoM/pull/19 (e.g., deploy.py) into [a separate module](https://github.com/OpenAdaptAI/OpenAdapter?tab=readme-ov-file) for importing, reducing boilerplate and keeping deployments consistent. Additionally, I am planning to extend it to support ECR/ECS for performance/scalability.

The new `deploy.py` here will be much shorter:
```
"""
OmniParser Deployment Script using OpenAdapter

This script automates deployment, setup, and management of the OmniParser server on AWS.

Setup Instructions:
    1. Create a `.env` file in the same directory with these keys:

        AWS_ACCESS_KEY_ID=<AWS access key>
        AWS_SECRET_ACCESS_KEY=<AWS secret key>
        AWS_REGION=<AWS region>
        GITHUB_OWNER=<GitHub org/user>
        GITHUB_REPO=<repo name>
        GITHUB_TOKEN=<GitHub token>
        PROJECT_NAME=OmniParser

    2. Install dependencies:
        pip install -r deploy_requirements.txt

Usage Examples:
    Start server:
        python deploy.py start

    Stop server and clean up resources:
        python deploy.py stop

    Check server status:
        python deploy.py status

    SSH into server:
        python deploy.py ssh
"""

from openadapter.server import OpenAdapterConfig, Deploy
import fire

config = OpenAdapterConfig(
    AWS_ACCESS_KEY_ID="",
    AWS_SECRET_ACCESS_KEY="",
    AWS_REGION="us-west-2",
    GITHUB_OWNER="OpenAdaptAI",
    GITHUB_REPO="OmniParser",
    GITHUB_TOKEN="",
    PROJECT_NAME="OmniParser"
)

class OmniParserDeploy(Deploy):
    def __init__(self):
        super().__init__(config=config)

if __name__ == "__main__":
    fire.Fire(OmniParserDeploy)
```

Let me know if you'd like to wait until this new module is ready before merging, or if you'd prefer proceeding now with the current structure. I suggest merging as-is; I will be more than happy to submit a new PR once the proposed module is available on PyPI.

Thanks again for your great work on this repo!

yadong-lu commented:
Hi @abrichr, really appreciate this pr!  since it contains a lot of changes, do you think we can keep the changes only for docker and aws deployment related. E.g. can we remove the change from gradio_demo.py? Thanks.

abrichr commented:
@yadong-lu I have reverted `gradio_demo.py` and added only the minimum required changes in order to return the bboxes and format the output in the text box with new lines:

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/a10ef52d-fec7-4fdc-81ba-1f99243dea14">

```
% python client.py http://3.238.21.15:7861  ~/oa/OpenAdapt/tests/assets/excel.png
Loaded as API: http://3.238.21.15:7861/ ✔
2024-11-01 11:34:48.401 | INFO     | __main__:predict_and_save:116 - Parsed content saved to: result_data_20241101_113437.json
2024-11-01 11:34:48.404 | INFO     | __main__:predict_and_save:122 - Output image saved to: output_image_20241101_113437.png
(venv) abrichr@MacBook-Pro-5 OmniParser % 
(venv) abrichr@MacBook-Pro-5 OmniParser % head result_data_20241101_113437.json
{
    "parsed_content_list": [
        "Text Box ID 0: OO0",
        "Text Box ID 1: A   9C .",
        "Text Box ID 2: Book1",
        "Text Box ID 3: Q",
        "Text Box ID 4: Home",
        "Text Box ID 5:  Insert",
        "Text Box ID 6: Draw",
        "Text Box ID 7: Page Layout",
(venv) abrichr@MacBook-Pro-5 OmniParser % tail result_data_20241101_113437.json
            0.02726871740120013
        ],
        "95": [
            0.9309752067440668,
            0.0796903748189267,
            0.0236164667536581,
            0.02922168473288605
        ]
    }
}%   
```

You can see the resulting successful Github action here: https://github.com/OpenAdaptAI/OmniParser/actions/runs/11631460372/job/32392578265

Ready for re-review.

n-sviridenko commented:
@abrichr I'd also add this to `download.py`:

```python
from utils import reader as easyocr_reader
print(f"Downloaded EasyOCR model")

from utils import paddle_ocr
print(f"Downloaded PaddleOCR model")
```

So that it doesn't spend GPUs on that

abrichr commented:
See https://github.com/OpenAdaptAI/OpenAdapt/pull/943 for an updated version that works independently of Github Actions and is compatible with the OmniParser v2 (weights and FastAPI server).
